engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java

```diff
@@ -180,70 +208,119 @@ private Void handleCreateTemplateFromSnapshot(SnapshotInfo snapshotInfo, Templat
+                createVolumeFromSnapshot(hostVO, snapshotInfo, true);
+            }
 
-            Map<String, String> srcDetails = getSnapshotDetails(_storagePoolDao.findById(srcDataStore.getId()), snapshotInfo);
+            DataStore srcDataStore = snapshotInfo.getDataStore();
 
-            copyCommand.setOptions(srcDetails);
+            String value = _configDao.getValue(Config.PrimaryStorageDownloadWait.toString());
+            int primaryStorageDownloadWait = NumbersUtil.parseInt(value, Integer.parseInt(Config.PrimaryStorageDownloadWait.getDefaultValue()));
+            CopyCommand copyCommand = new CopyCommand(snapshotInfo.getTO(), templateInfo.getTO(), primaryStorageDownloadWait, VirtualMachineManager.ExecuteInSequence.value());
+
+            String errMsg = null;
+
+            CopyCmdAnswer copyCmdAnswer = null;
 
-            copyCmdAnswer = (CopyCmdAnswer)_agentMgr.send(hostVO.getId(), copyCommand);
-        }
-        catch (Exception ex) {
-            throw new CloudRuntimeException(ex.getMessage());
-        }
-        finally {
             try {
-                _volumeService.revokeAccess(snapshotInfo, hostVO, srcDataStore);
+                // If we are using a back-end snapshot, then we should still have access to it from the hosts in the cluster that hostVO is in
+                // (because we passed in true as the third parameter to createVolumeFromSnapshot above).
+                if (usingBackendSnapshot == false) {
+                    _volumeService.grantAccess(snapshotInfo, hostVO, srcDataStore);
+                }
+
+                Map<String, String> srcDetails = getSnapshotDetails(snapshotInfo);
+
+                copyCommand.setOptions(srcDetails);
+
+                copyCmdAnswer = (CopyCmdAnswer)_agentMgr.send(hostVO.getId(), copyCommand);
             }
             catch (Exception ex) {
-                s_logger.debug(ex.getMessage(), ex);
+                throw new CloudRuntimeException(ex.getMessage());
             }
-
-            if (copyCmdAnswer == null || !copyCmdAnswer.getResult()) {
-                if (copyCmdAnswer != null && copyCmdAnswer.getDetails() != null && !copyCmdAnswer.getDetails().isEmpty()) {
-                    errMsg = copyCmdAnswer.getDetails();
+            finally {
+                try {
+                    _volumeService.revokeAccess(snapshotInfo, hostVO, srcDataStore);
                 }
-                else {
-                    errMsg = "Unable to perform host-side operation";
+                catch (Exception ex) {
+                    s_logger.debug(ex.getMessage(), ex);
                 }
-            }
 
-            try {
-                if (errMsg == null) {
-                    snapshotInfo.processEvent(Event.OperationSuccessed);
+                if (copyCmdAnswer == null || !copyCmdAnswer.getResult()) {
+                    if (copyCmdAnswer != null && copyCmdAnswer.getDetails() != null && !copyCmdAnswer.getDetails().isEmpty()) {
```

Comment:
@mike-tutkowski Could you please use StringUtils.isEmpty (https://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/StringUtils.html#isEmpty(java.lang.String))?

This is a null safe test if the String **copyCmdAnswer.getDetails** is empty (you can use this method at lines 260, 363 and 403 -- in the **StorageSystemDataMotionStrategy** class).

Thanks.